xdsl/builder.py

```diff
@@ -12,47 +12,92 @@
 
 
 @dataclass
-class Builder:
+class InsertPoint:
     """
-    A helper class to construct IRs, by keeping track of where to insert an
-    operation. Currently the insertion point is always at the end of the block.
-    In the future will closely follow the API of `OpBuilder` in MLIR, inserting
-    at arbitrary locations.
+    An insert point.
+    It is either a point before an operation, or at the end of a block.
 
-    https://mlir.llvm.org/doxygen/classmlir_1_1OpBuilder.html
+    https://mlir.llvm.org/doxygen/classmlir_1_1OpBuilder_1_1InsertPoint.html
     """
 
     block: Block
+    """The block where the insertion point is in."""
+
+    op_before: Operation | None = field(default=None)
     """
-    Operations will be inserted in this block.
+    The operation right before the insertion point.
+    None if the insertion point is at the end of the block.
     """
 
-    _insertion_point: Operation | None = field(default=None)
+    @overload
+    def __init__(self, point: Operation) -> None:
+        ...
+
+    @overload
+    def __init__(self, point: Block) -> None:
+        ...
+
+    @overload
+    def __init__(self, point: Block, point2: Operation | None) -> None:
+        ...
+
+    def __init__(
+        self, point: Block | Operation, point2: Operation | None = None
+    ) -> None:
```

Comment:
Could this be `{inner,outer}_point` or something? `point{1,2}` seems like suboptimal naming :wink:

Reply:
I have a feeling that it can't be due to Python's naming things, but I would suggest `op_before_or_block, op_before`